src/hotspot/share/cds/filemap.cpp

```diff
@@ -1069,117 +1075,185 @@ class FileHeaderHelper {
+      return false;
+    }
+
+    size_t filelen = os::lseek(fd, 0, SEEK_END);
+    if (gen_header._header_size >= filelen) {
+      FileMapInfo::fail_continue("Archive file header larger than archive file");
+      return false;
+    }
+
+    // Read the actual header and perform more checks
+    size = gen_header._header_size;
+    _header = (GenericCDSFileMapHeader*)NEW_C_HEAP_ARRAY(char, size, mtInternal);
+    lseek(fd, 0, SEEK_SET);
+    n = os::read(fd, (void*)_header, (unsigned int)size);
+    if (n != size) {
+      FileMapInfo::fail_continue("Unable to read actual CDS file map header from shared archive");
+      return false;
+    }
+
+    if (!check_crc()) {
+      return false;
+    }
+
+    if (!check_and_init_base_archive_name()) {
+      return false;
+    }
+
+    // All fields in the GenericCDSFileMapHeader has been validated.
+    _is_valid = true;
     return true;
   }
 
   GenericCDSFileMapHeader* get_generic_file_header() {
-    return &_header;
-  }
-
-  char* read_base_archive_name() {
-    assert(_fd != -1, "Archive should be open");
-    size_t name_size = _header._base_archive_name_size;
-    assert(name_size != 0, "For non-default base archive, name size should be non-zero!");
-    char* base_name = NEW_C_HEAP_ARRAY(char, name_size, mtInternal);
-    lseek(_fd, _header._base_archive_name_offset, SEEK_SET); // position to correct offset.
-    size_t n = os::read(_fd, base_name, (unsigned int)name_size);
-    if (n != name_size) {
-      log_info(cds)("Unable to read base archive name from archive");
-      FREE_C_HEAP_ARRAY(char, base_name);
-      return nullptr;
+    assert(_header != nullptr && _is_valid, "must be a valid archive file");
+    return _header;
+  }
+
+  const char* base_archive_name() {
+    assert(_header != nullptr && _is_valid, "must be a valid archive file");
+    return _base_archive_name;
+  }
+
+ private:
+  bool check_crc() {
+    if (VerifySharedSpaces) {
+      FileMapHeader* header = (FileMapHeader*)_header;
+      int actual_crc = header->compute_crc();
+      if (actual_crc != header->crc()) {
+        log_info(cds)("_crc expected: %d", header->crc());
+        log_info(cds)("       actual: %d", actual_crc);
+        FileMapInfo::fail_continue("Header checksum verification failed.");
+        return false;
+      }
     }
-    if (base_name[name_size - 1] != '\0' || strlen(base_name) != name_size - 1) {
-      log_info(cds)("Base archive name is damaged");
-      FREE_C_HEAP_ARRAY(char, base_name);
-      return nullptr;
+    return true;
+  }
+
+  bool check_and_init_base_archive_name() {
+    unsigned int name_offset = _header->_base_archive_name_offset;
+    unsigned int name_size   = _header->_base_archive_name_size;
+    unsigned int header_size = _header->_header_size;
+
+    if (name_offset + name_size < name_offset) {
+      FileMapInfo::fail_continue("base_archive_name/size overflow: " UINT32_FORMAT "/" UINT32_FORMAT,
+                                 name_offset, name_size);
```

Comment:
Since you're printing the `name_offset` and `name_size`, maybe it is clearer to change the message to:

"base_archive_name offset/size overflow: ..."